### PR TITLE
Create router errors

### PR DESCRIPTION
**Pull request by lmichelin**
This PR improves the way router errors are thrown.

4 router errors types have been created:

- `NavigationDuplicated`
- `NavigationCancelled`
- `NavigationRedirected`
- `NavigationAborted`

Before:

```
Uncaught (in promise) undefined
```

After:

```
Uncaught (in promise) Error: NavigationAborted: Navigating to location ("/foo?abort=y") has been aborted
    at eval (base.js:157)
    at eval (app.js:33)
    at iterator (base.js:153)
    at step (async.js:13)
    at step (async.js:17)
    at runQueue (async.js:21)
    at HashHistory.confirmTransition (base.js:179)
    at HashHistory.transitionTo (base.js:74)
```

### CODE REVIEW

**Comment by lmichelin** [comment]
@posva I wrote some unit tests, the PR is ready to be reviewed 😉

**Comment by ariabuckles** [comment]
Bump on this? This would be really helpful for tracking down where promise rejections are coming from <3

**Comment by bponomarenko** [comment]
It seems this PR can also solve #3027. +1 for it to be merged and released.

**Comment by quentinus95** [comment]
@lmichelin could it be possible to export the exceptions so they can be imported (e.g. using `import { NavigationDuplicated } from "vue-router"`)? It would allow catching the exceptions and use `instanceof` (especially when coding with TypeScript)  to take decisions based on the thrown exceptions.

**Comment by posva** [comment]
@quentinus95 `instanceof` with extended Errors doesn't work on some browsers like IE.
This PR is on hold until https://github.com/vuejs/vue-router-next/pull/123 is merged to make sure exported errors share the same naming and reduce migration cost

**Review by posva — CHANGES_REQUESTED** [review]
Hi @lmichelin 

Now that errors are stable on vue-router-next, I think we could try to align the user API part to make migration easier.
Namely, we need:

- a `type`, `from` property
- an object (to simulate the enumeration) with the same properties as [`NavigationFailureType`](https://github.com/vuejs/vue-router-next/blob/master/src/errors.ts#L26-L30)
- an exported TS interface so users can type errors

We can also remove:

- the `name` and `_name` properties
- all the code to hack around Error extension and use plain errors (https://github.com/vuejs/vue-router-next/blob/master/src/errors.ts#L77-L90)

It's okay to have a few differences like the lack of `from` and `to` properties. Also, let me know if you don't have the time so I can pick it up. If you only manage to add a few things but not all, that's, of course, fine as well!

Link to RFC: https://github.com/vuejs/rfcs/pull/150
Link to vue-router-next errors file: https://github.com/vuejs/vue-router-next/blob/master/src/errors.ts

File: package.json
Comment (posva) [inline package.json]: Let's keep this change out of the PR. If needed I will upgrade this on dev and you can then merge or rebase against it

File: src/history/abstract.js
Comment (posva) [inline src/history/abstract.js]: We probably want to keep the check of duplicated navigation

File: src/history/base.js
Comment (posva) [inline src/history/base.js]: same as the comment above, I think we should keep the check of the error type

File: src/history/errors.js
Comment (posva) [inline src/history/errors.js]: let's keep the same names as the public API: https://github.com/vuejs/vue-router-next/blob/master/src/errors.ts#L26-L30 (`NavigationFailureType`)

File: src/history/errors.js
Comment (posva) [inline src/history/errors.js]: Let's take the same error messages as in https://github.com/vuejs/vue-router-next/blob/master/src/errors.ts#L47-L72

File: src/history/errors.js
Comment (posva) [inline src/history/errors.js]: this is probably a leftover

**Review by posva — CHANGES_REQUESTED** [review]

File: package.json
Comment (lmichelin) [reply package.json]: I did this change to be able to run e2e tests. Without this update, e2e tests can't run: https://circleci.com/gh/vuejs/vue-router/5121 😉

Comment (posva) [reply package.json]: oh wow, they are still green, that's so wrong lol
I updated on `dev` the chromedriver but there are a few tests failing I need to fix. After that you can merge dev into this (don't worry about the git commit history, we squash at the end)

Comment (posva) [reply package.json]: It's good now!

File: src/util/warn.js
Comment (posva) [inline src/util/warn.js]: I think we can make this simpler:

```suggestion
  return isError(err) && err._isRouter && (errorType == null || err.type === errorType)
```

File: src/history/base.js
Comment (posva) [inline src/history/base.js:113]: We can remove the `isError` check since `isRouterError` already checks that
```suggestion
      if (!isRouterError(err, NavigationFailureType.NAVIGATION_DUPLICATED)) {
```

File: src/history/errors.js
Comment (posva) [inline src/history/errors.js]: The values do not matter much since they shouldn't be used directly but I think it's better to force that behavior by using numbers (like an enum), to prevent the mistake of people using a simple string comparison: `error.type === 'NavigationRedirected'` instead of `error.type === NavigationFailureType.redirected`

We should also keep the same keys as in https://github.com/vuejs/vue-router-next/blob/master/src/errors.ts#L27-L29

```suggestion
  redirected: 1,
  aborted: 2,
  cancelled: 3,
  duplicated: 4
```

**Review by posva — CHANGES_REQUESTED** [review]
I realised you have a sample for an e2e test but it doesn't seem like you need it since you added a unit test already

**Comment by lmichelin** [comment]
> I realised you have a sample for an e2e test but it doesn't seem like you need it since you added a unit test already

`examples/router-errors/app.js` is an example to show how to throw the errors, it's not made for e2e purposes, but just to show simple examples for these errors.

File: src/history/base.js
Comment (lmichelin) [reply src/history/base.js:113]: I don't think it's equivalent: If I understood correctly we want to check if `err` is an error but not a navigation duplicated error. When `err` is `null` (like here https://github.com/vuejs/vue-router/blob/588220c475b29e038bc6c0badddac7992bf27445/src/history/base.js#L188) the result of the condition will not be the same.

Comment (posva) [reply src/history/base.js:113]: You are right,  misread the line

**Review by posva — APPROVED** [review]
Thanks a lot for your help with this @lmichelin and sorry for the delay!

File: src/history/base.js
Comment (schmidtk) [inline src/history/base.js:113]: This should be renamed to `NavigationFailureType.duplicated`.

File: src/history/abstract.js
Comment (schmidtk) [inline src/history/abstract.js:54]: Here as well.

**Review by schmidtk — COMMENTED** [review]
Good morning @lmichelin and @posva. I discovered this PR while handling navigation duplicated errors in my application and have been tracking the changes in case it requires migration when we update `vue-router` in the future.

I noted two places in the changes where `NavigationFailureType.NAVIGATION_DUPLICATED` references were not updated after the keys were changed in `NavigationFailureType`. I'm admittedly new to `vue-router` internals, so apologies if I misunderstood something here. Thanks for your time.

**Comment by lmichelin** [comment]
You are right, I completely forgot these references 🙈

**Comment by lmichelin** [comment]
I just opened a new PR to fix this: https://github.com/vuejs/vue-router/pull/3204

**Comment by alecgibson** [comment]
I think this was a breaking change? We don't get the `name` property on the `Error` any more, which is useful when trying to handle particular errors (eg ignoring `NavigationDuplicated`). At the very least it may be nice to expose `isRouterError` as a utility method? And also to expose the failure types, especially because it's an opaque enum.

**Comment by posva** [comment]
The errors were not documented and considered private. Checking navigation failures docs are on the way!

**Comment by alecgibson** [comment]
I'd say that checking an `Error.name` field is pretty [standard practice](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Error/name), especially if it's not been prefixed with an underscore. I don't mind a change, but I'm just saying this could be considered breaking.

**Comment by snoozbuster** [comment]
+1 for this being a breaking change - once again I am back here to figure out why suddenly there are errors in my app where there weren't errors before (and the fact that it was merged and published without docs on how to do proper error checking also seems like a mistake)

**Comment by snoozbuster** [comment]
@posva - are these type constants exported from the package, so that I can inspect navigation errors in my application and handle them appropriately? I don't see anything which adds that, is it in another PR?

**Comment by afv** [comment]
I was checking `err.name === 'NavigationDuplicated'` before but after these changes the name is just `'Error'`.

As a workaround I'm now importing `NavigationFailureType` at `main.js` and adding it to `window`:

```
import { NavigationFailureType } from 'vue-router/src/history/errors';

window.NavigationFailureType = NavigationFailureType;
```

On components I can now check `err.type === window.NavigationFailureType.duplicated`.

**Comment by posva** [comment]
Please check https://github.com/vuejs/vue-router/pull/3220 for the proposed documentation for errors. Your feedback could be valuable there

**Pull request merged**
